### DISCUSSION TITLE
Help Wanted for Mega2560

### DISCUSSION THREAD

**Discussion by basitalikhan** [JAndrassy/ArduinoOTA#67] [General]
Can you guide or give any info about implementing it over any GSM module (SIM800 series/Telit GSM Modules) through GPRS? Maybe in TCP Transparent Mode.

**Comment by JAndrassy** [comment]
I would use the TinyGSM library, which has a standard Client. but I never used a GSM module or the library. 
see the OTASketchDownload example

**Reply by basitalikhan** [reply]
Thank you. I used the TinyGSM library but the http client example there donot download the whole file so I am manually reading the file using

AT+HTTPACTION=0
AT+HTTPREAD
commands and I can read the whole file but then if I pass those bytes to

InternalStorage.write(b); 

the microcontroller immediately restarts.  I am using Arduino mega 2560 while I am uploading code through USBASP. The bin file that I am downloading is generated by arduino in the temp folder.

The Code that I am uploading to ATMEGA2560 is with the bootloader OTA_Test.ino.with_bootloader.hex so I can say that the code contains the bootloader.

`## Arduino ATmega2560 Arduino OTA Test
## -------------------------
mega.menu.cpu.atmega2560OTA=ATmega2560 (OTA Test Version)

mega.menu.cpu.atmega2560OTA.upload.protocol=wiring
mega.menu.cpu.atmega2560OTA.upload.maximum_size=258048
mega.menu.cpu.atmega2560OTA.upload.speed=115200

mega.menu.cpu.atmega2560OTA.bootloader.high_fuses=0xD9
mega.menu.cpu.atmega2560OTA.bootloader.extended_fuses=0xFD
mega.menu.cpu.atmega2560OTA.bootloader.file=optiboot8/optiboot_atmega2560.hex

mega.menu.cpu.atmega2560OTA.build.mcu=atmega2560
mega.menu.cpu.atmega2560OTA.build.board=AVR_MEGA2560
`

I have used the bootloader from your my_boards repo optiboot8

**Reply by JAndrassy** [reply]
but the bin which you download over GSM module is a normal bin without bootloader?

**Reply by basitalikhan** [reply]
It is with the bootloader

**Reply by JAndrassy** [reply]
how should that work? what is the size of that bin?

**Reply by basitalikhan** [reply]
The bin file is 64KB. The sketch only contains the blink code and nothing else.

**Reply by JAndrassy** [reply]
I still try to understand why do you use the bin with bootloader? because
1) it is wrong. there is a bug for Mega. the bin with bootloader has to be 256 kB because the bootloader is at the end of the flash and the sketch at the beginning (hex file contains target addresses so it can skip empty location, bin can't)
2) 256 kB would be of course more then half of the flash so there is no place to store it
3) the bootloader does the copying of the uploaded bin so it can't override itself

**Reply by basitalikhan** [reply]
Actually I think arduino does not create BIN file without bootloader. I overlooked it. Is there any way to compile the BIN file without bootloader or use the HEX file and convert it to BIN file using online tool.

**Reply by JAndrassy** [reply]
I have in console of Arduino IDE (with my_boards board selected)
```
/home/duro/arduino-1.8.13/hardware/tools/avr/bin/avr-objcopy -O binary /tmp/arduino_build_602381/ASCIITable.ino.elf /tmp/arduino_build_602381/ASCIITable.ino.bin
/home/duro/arduino-1.8.13/hardware/tools/avr/bin/avr-objcopy -O ihex -j .eeprom --set-section-flags=.eeprom=alloc,load --no-change-warnings --change-section-lma .eeprom=0 /tmp/arduino_build_602381/ASCIITable.ino.elf /tmp/arduino_build_602381/ASCIITable.ino.eep
/home/duro/arduino-1.8.13/hardware/tools/avr/bin/avr-objcopy -O ihex -R .eeprom /tmp/arduino_build_602381/ASCIITable.ino.elf /tmp/arduino_build_602381/ASCIITable.ino.hex
```

**Reply by JAndrassy** [reply]
if you add to platform.txt in my_boards/avr
```
build.preferred_out_format=bin

## Save hex
recipe.output.tmp_file={build.project_name}.{build.preferred_out_format}
recipe.output.save_file={build.project_name}.{build.variant}.{build.preferred_out_format}
```
it will export .bin on "Export compiled binary" in Tools menu.

**Reply by basitalikhan** [reply]
Thankyou. I think now the BIN file generated is good, the size is 1,534 bytes. But as soon as it starts to download file in 
```
while (length > 0) {
    if (!Serial1.readBytes(&b, 1)) // reading a byte with timeout
      break;
    InternalStorage.write(b);
    length--;
  }
```
  
the controller restarts. Anything else that I am missing? Are you on any freelancer website as I want it to complete it soon?

**Reply by JAndrassy** [reply]
shouldn't you read from client object?

**Reply by basitalikhan** [reply]
I am reading it from serial because if I use client object to download the file it does not download the file completely. So I use the 

```
AT+HTTPPARA=
AT+HTTPACTION=0
AT+HTTPREAD
```

commands to read the whole file and it does print the contents of the whole file correctly.

**Reply by basitalikhan** [reply]
Thank you. This time I used another site for the file and used your method and I did OTA successfully. Is this OTA functionality available on ATMega328P or ATMega328PB? What does optiboot_atmega328_big.hex contains?

**Reply by JAndrassy** [reply]
it works on ATmega328p if your sketch is smaller then half of the available flash (the bootloader takes 1 kB). 
 optiboot_atmega328_big.hex has the copy_flash_pages function for ArduinoOTA

**Reply by basitalikhan** [reply]
The same code does not compile if I select the Uno BigBoot device.

`OTA_Test_3:206:10: error: 'InternalStorage' was not declared in this scope`

**Reply by JAndrassy** [reply]
yes it is suppressed with ifdef on memory size. in ArduinoOTA.h and in InternalStorageAVR.cpp and h

**Reply by basitalikhan** [reply]
For Atmega2560 InternalStorageAVR is used or simply InternalStorage.

because in the main code I am simply using InternalStorage.write(b) commands so which class does it use? The code is running fine with InternalStorage.write(b)

**Reply by JAndrassy** [reply]
object's name is always InternalStorage

**Reply by basitalikhan** [reply]
Actually I want to use this library on Atmel Studio so I will convert it to C language for ATmega2560 that is why I was asking. Because your ArduinoOTA.h is using:

```
#include "WiFiOTA.h"
#ifdef __AVR__
#if FLASHEND >= 0xFFFF
#include "InternalStorageAVR.h"
#endif
#elif defined(ESP8266) || defined(ESP32)
#include "InternalStorageESP.h"
#else
#include "InternalStorage.h"
#endif
#ifdef __SD_H__
#include "SDStorage.h"
SDStorageClass SDStorage;
#endif
#ifdef SerialFlash_h_
#include "SerialFlashStorage.h"
SerialFlashStorageClass SerialFlashStorage;
#endif
```

So I was searching which file it is using.
